### ISSUE TITLE
SSH from WSL

### ISSUE THREAD

**Issue by LukeIreland1** [microsoft/WSL#5017]
When connected to my company's VPN, I can ssh from cmd fine. I want to be able to ssh from my WSL zsh as well, but I get `Could not resolve hostname <hostname>: Name or service not known`

**Comment by crramirez** [comment]
Ok give us more data. 

1. You are using WSL1 or WSL2? 
2. When you issue `cat /etc/resolv.conf` which is the result?
3. This error message sounds strange to me "hostname" how are you invoking SSH? which is the command?

Regards,
Carlos

**Comment by LukeIreland1** [comment]
1) WSL1, but I thought I was on 2. I've just checked which now, and apparently I'm on 1, so I will try with 2.
2)```# This file was automatically generated by WSL. To stop automatic generation of this file, remove this line.
nameserver 192.168.0.1
nameserver fdf7:8e43:682f:0:3e89:94ff:fefd:218
nameserver 10.1.2.23
search Home arm.com```
3)ssh lukire01@taishan-3.on.arm.com

This works fine from my MacBook when connected to the VPN, and fine from cmd when connected to the VPN.

**Comment by crramirez** [comment]
Stay in WSL1 it is better with network issues like this. It is in fact a DNS problem. 

I advise you to use `ssh -vvv lukire01@taishan-3.on.arm.com` to see what ssh is doing.

**Comment by LukeIreland1** [comment]
```luke@Hyperion:~$ ssh -vvv lukire01@taishan-3.on.arm.com
OpenSSH_7.6p1 Ubuntu-4ubuntu0.3, OpenSSL 1.0.2n  7 Dec 2017
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: Applying options for *
debug2: resolving "taishan-3.on.arm.com" port 22
ssh: Could not resolve hostname taishan-3.on.arm.com: Name or service not known

**Comment by crramirez** [comment]
Well as a workaround edit your Windows Hosts file and this domain and its IP. I use this tool to edit it: https://hostsfileeditor.com/

But for any reason, the DNS that comes from your VPN isn't going into /etc/resolv.conf

**Comment by LukeIreland1** [comment]
If I can find that DNS, how would I put it into /etc/resolv.conf? As in, what would an example configuration look like? Is it just nameserver and the DNS?

I think I understand your Windows hosts instructions, so I'll give that a go also.

**Comment by crramirez** [comment]
Yes you can but will be overwritten on the next launch.

**Comment by bbulkow** [comment]
Interesting that I raised an issue just now very similar in that the resolv.conf becomes incorrect if there is an IPv6 in the resolve list. One sses in this example that the 10. , which is almost certainly the VPN resolver, is at the end, and there is an IPv6 in the middle. I was on WSL1 for sure. So maybe this is an issue in both?

Workaround is hard to apply. I am on and off these VPNs all the time.

**Comment by liv3000** [comment]
Is there anything new on this topic as I am having the same issue with the WSL and ssh'ing into my uni network (with their vpn on)

**Comment by vineethelias** [comment]
1. Check if you can ssh using IP
2. Do a debug with ssh -vvv
3. Check if you can resolve the DNS name with nslookup
4. Make sure the nssswitch.conf is good.

If you still not able to get is work, follow below steps.

Automated steps: https://gist.github.com/matthiassb/9c8162d2564777a70e3ae3cbee7d2e95

First run on powershell:
Get-DnsClientServerAddress -AddressFamily IPv4 | Select-Object -ExpandPropert ServerAddresses

Then run bash on windows with either ubuntu or bash commands (depending on your installation). Edit the /etc/resolv.conf and replace all nameserver {ip} with the ones generated by the top one.

You should have internet and intranet back now. Remove the first line to make it permanent but be aware if you change networks it will not update anymore. (back it up first maybe?)

**Comment by Noexpert** [comment]
I have the same issue:
`ssh: Could not resolve hostname hostname.com: Temporary failure in name resolution`

Possibly related to https://github.com/microsoft/WSL/issues/4992

**Comment by AymenFJA** [comment]
I did a workaround by using the default windows `OpenSSH` client instead of WSL one: 
1- Locate your OpenSSH on Windows by doing:
```
$ where ssh
C:\Windows\System32\OpenSSH\ssh.exe
```

2- From your WSL terminal ssh to your target machine by using the located OpenSSH client from above: 
```
/mnt/c/Windows/System32/OpenSSH/ssh.exe -i /home/user/.ssh/private_key  name@ip
```

**Comment by AQZ0216** [comment]
> As a temporary workaround by using the default windows `OpenSSH` client instead of WSL one: 1- Open a `Command Prompt terminal` and locate your OpenSSH on Windows by doing:
> 
> ```
> $ where ssh
> C:\Windows\System32\OpenSSH\ssh.exe
> ```
> 
> 2- From your `WSL terminal` ssh to your target machine by using the located OpenSSH client from above:
> 
> ```
> /mnt/c/Windows/System32/OpenSSH/ssh.exe -i /home/user/.ssh/private_key  name@ip
> ```
> 
> FYI: similar to `ssh` your can use `scp` if you want to upload/download data.

this works

**Comment by brycepg** [comment]
Here's a hack to get git working inside of wsl with @NDR0216 suggestion:

    sudo mv /usr/bin/ssh /usr/bin/ssh.broken
    sudo ln -s /mnt/c/Windows/System32/OpenSSH/ssh.exe /usr/bin/ssh

**Comment by brycepg** [comment]
@kislow's solution works for the package manager too 😽 

### TLDR

Just replace the WSL `/etc/resolv.conf` with:
    
    [network]
    generateResolvConf = false
    nameserver 8.8.8.8

**Comment by pmartincic** [comment]
I believe this was fixed by DNS tunnelling. https://learn.microsoft.com/en-us/windows/wsl/networking#dns-tunneling
